File: 1026-maximum-difference-between-node-and-ancestor/1026-maximum-difference-between-node-and-ancestor.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int preOrder(TreeNode root, int minimum, int maximum){
        if(root == null){
            return maximum - minimum;
        }
        maximum = Math.max(maximum, root.val);
        minimum = Math.min(minimum, root.val);
        
        return Math.max(preOrder(root.left, minimum, maximum), preOrder(root.right, minimum, maximum));
        
    }
    public int maxAncestorDiff(TreeNode root) {
        return preOrder(root,99999,-99999);
        
    }
}

File: 1026-maximum-difference-between-node-and-ancestor/1026-maximum-difference-between-node-and-ancestor.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def traverse(self, root: Optional[TreeNode], min_value: int, max_value: int) -> int:
        if root is None:
            return max_value - min_value
        
        min_value = min(min_value, root.val)
        max_value = max(max_value, root.val)
        
        return max(self.traverse(root.left, min_value, max_value), 
                   self.traverse(root.right, min_value, max_value))
                   
    def maxAncestorDiff(self, root: Optional[TreeNode]) -> int:
        return self.traverse(root, 99999, -99999)
        